test/05-updates.test.js

```diff
@@ -27,57 +27,61 @@ test('Pear.updates(listener) should notify when restaging and releasing applicat
+  ok(versions?.app, 'updater is ready')
+
+  const untilUpdate1 = Helper.untilResult(pipe, 30_000).then((data) => JSON.parse(data))
+  const untilUpdate2 = untilUpdate1.then(() => Helper.untilResult(pipe, 30_000)).then((data) => JSON.parse(data))
 
   comment('2. Create new file, restage, and reseed')
 
   const file = `${ts()}.tmp`
   comment(`\tcreating test file (${file})`)
-  fs.writeFileSync(path.join(harness, file), 'test')
-  teardown(() => { try { fs.unlinkSync(path.join(harness, file)) } catch { /* ignore */ } })
+  fs.writeFileSync(path.join(updates, file), 'test')
+  teardown(() => { try { fs.unlinkSync(path.join(updates, file)) } catch { /* ignore */ } })
+
   comment('\tstaging')
   const stager2 = new Helper(rig)
   teardown(() => stager2.close(), { order: Infinity })
   await stager2.ready()
 
-  await Helper.pick(stager2.stage(stageOpts(testId)), { tag: 'final' })
+  const staging2 = stager2.stage(stageOpts(testId))
+  teardown(() => Helper.teardownStream(staging2))
+  await Helper.pick(staging2, { tag: 'final' })
 
-  fs.unlinkSync(path.join(harness, file))
+  fs.unlinkSync(path.join(updates, file))
 
-  const update1 = await update1ActualPromise
-  const update1Version = update1?.value?.version
+  const update1 = await untilUpdate1
+  const update1Version = update1?.version
   is(hypercoreid.normalize(update1Version?.key), hypercoreid.normalize(key), 'app updated with matching key')
   is(update1Version?.fork, 0, 'app version.fork is 0')
   ok(update1Version?.length > 0, `app version.length is non-zero (v${update1Version?.fork}.${update1Version?.length})`)
 
-  comment('releasing')
-  const update2Promise = await update2LazyPromise
-  const update2ActualPromise = running.inspector.awaitPromise(update2Promise.objectId)
+  comment('\treleasing')
   const releaser = new Helper(rig)
-  teardown(() => releaser.close())
+  teardown(() => releaser.close(), { order: Infinity })
   await releaser.ready()
 
   const releasing = releaser.release(releaseOpts(testId, key))
+  teardown(() => Helper.teardownStream(releasing))
   await Helper.pick(releasing, { tag: 'released' })
-  teardown(() => releaser.close()) // TODO why is this needed?
-  comment('waiting for update')
-  const update2 = await update2ActualPromise
-  const update2Version = update2?.value?.version
+
+  comment('\twaiting for update')
+  const update2 = await untilUpdate2
+  const update2Version = update2?.version
   is(hypercoreid.normalize(update2Version?.key), hypercoreid.normalize(key), 'app updated with matching key')
   is(update2Version?.fork, 0, 'app version.fork is 0')
   ok(update2Version?.length > update1Version?.length, `app version.length incremented (v${update2Version?.fork}.${update2Version?.length})`)
-  await running.inspector.evaluate('__PEAR_TEST__.close()')
-  await running.inspector.close()
-  const { code } = await running.until.exit
-  is(code, 0, 'exit code is 0')
+
+  pipe.write('exit')
```

Comment:
can we try pipe.end() for parent?

then in the child worker, we can pipe.on('end', cb) or pipe.on('close', cb) with a cb to updates.end()

Reply:
@maidh91 Hmm.. tried this but it didn't emit the end nor the close event on the worker.

Reply:
@rafapaezbas Hmm.. that one seems to be using Pear.teardown() instead of a pipe.on('end', ...) handler on the worker side.

Pear.teardown() won't trigger in the updates worker because Pear.updates() is still listening.

Reply:
Turns out we need to resume() the pipe on the worker side for end to be emitted correctly to switch its stream to flowing mode. Discussed in slack.


https://nodejs.org/api/stream.html#event-end
> The 'end' event will not be emitted unless the data is completely consumed. This can be accomplished by switching the stream into flowing mode, or by calling [stream.read()](https://nodejs.org/api/stream.html#readablereadsize) repeatedly until all data has been consumed.

Reply:
great finding, good to know that, thanks @jkcdarunday